File: JAVA/1106. Parsing A Boolean Expression.java
class Solution {
    int i=0;
    public boolean parseBoolExpr(String e) {
        return p(e);
    }
    private boolean p(String e){
        if(e.charAt(i)=='t'){
            i++;
            return true;
        }
        if(e.charAt(i)=='f'){
            i++;
            return false;
        }
        if(e.charAt(i)=='!'){
            i+=2;
            boolean a=!p(e);
            i++;
            return a;
        }
        boolean is=e.charAt(i)=='&';
        boolean a=is;
        i+=2;
        while(e.charAt(i)!=')'){
            boolean pa=p(e);
            if(is)
            a&=pa;
            else
            a|=pa;
            if(e.charAt(i)==',')
            i++;
        }
        i++;
        return a;
    }
}

File: C++/1106. Parsing A Boolean Expression.cpp
class Solution {
public:
int i=0;
    bool p(string e){
        if(e[i]=='t'){
            i++;
            return true;
        }
        if(e[i]=='f'){
            i++;
            return false;
        }
        if(e[i]=='!'){
            i+=2;
            bool a=!p(e);
            i++;
            return a;
        }
        bool is=e[i]=='&';
        bool a=is;
        i+=2;
        while(e[i]!=')'){
            bool pa=p(e);
            if(is)
            a&=pa;
            else
            a|=pa;
            if(e[i]==',')
            i++;
        }
        i++;
        return a;
    }
    bool parseBoolExpr(string e) {
        return p(e);
    }
};